File: pymab/policies/bayesian_ucb.py
"""Bayesian UCB policies."""

from __future__ import annotations

from statistics import NormalDist

import numpy as np

from pymab.policies.policy import ActionValuePolicy, FloatArray, Policy


class BernoulliBayesianUCBPolicy(ActionValuePolicy):
    """Bayesian UCB using Beta posterior quantiles."""

    def __init__(
        self,
        *,
        n_arms: int | None = None,
        n_bandits: int | None = None,
        alpha_prior: float = 1.0,
        beta_prior: float = 1.0,
        quantile: float = 0.95,
        **_: object,
    ) -> None:
        if alpha_prior <= 0 or beta_prior <= 0:
            raise ValueError("alpha_prior and beta_prior must be positive")
        if not 0 < quantile < 1:
            raise ValueError("quantile must be in (0, 1)")
        self.alpha_prior = float(alpha_prior)
        self.beta_prior = float(beta_prior)
        self.quantile = float(quantile)
        self.successes: FloatArray
        self.failures: FloatArray
        arms = n_arms if n_arms is not None else n_bandits
        if arms is None:
            raise TypeError("n_arms is required")
        super().__init__(n_arms=int(arms), initial_value=0.0)

    def reset(self) -> None:
        super().reset()
        self.successes = np.zeros(self.n_arms, dtype=float)
        self.failures = np.zeros(self.n_arms, dtype=float)

    def select_action(self, *, rng: np.random.Generator) -> int:
        # NumPy lacks a beta PPF. A large deterministic posterior sample is a
        # practical approximation for the quantile and keeps SciPy optional.
        samples = rng.beta(
            self.alpha_prior + self.successes[:, np.newaxis],
            self.beta_prior + self.failures[:, np.newaxis],
            size=(self.n_arms, 256),
        )
        upper_bounds = np.quantile(samples, self.quantile, axis=1)
        return int(np.argmax(upper_bounds))

    def update(self, *, action: int, reward: float) -> None:
        if reward not in (0, 1, 0.0, 1.0):
            raise ValueError("Bernoulli Bayesian UCB requires binary rewards")
        super().update(action=action, reward=float(reward))
        if reward > 0:
            self.successes[action] += 1.0
        else:
            self.failures[action] += 1.0

    def __repr__(self) -> str:
        return (
            "BernoulliBayesianUCBPolicy("
            f"quantile={self.quantile}, alpha_prior={self.alpha_prior}, "
            f"beta_prior={self.beta_prior})"
        )


class GaussianBayesianUCBPolicy(ActionValuePolicy):
    """Bayesian UCB for Gaussian rewards with known observation variance."""

    def __init__(
        self,
        *,
        n_arms: int | None = None,
        n_bandits: int | None = None,
        prior_mean: float = 0.0,
        prior_precision: float = 1.0,
        reward_precision: float = 1.0,
        quantile: float = 0.95,
        **_: object,
    ) -> None:
        if prior_precision <= 0 or reward_precision <= 0:
            raise ValueError("precisions must be positive")
        if not 0 < quantile < 1:
            raise ValueError("quantile must be in (0, 1)")
        self.prior_mean = float(prior_mean)
        self.prior_precision = float(prior_precision)
        self.reward_precision = float(reward_precision)
        self.quantile = float(quantile)
        self.means: FloatArray
        self.precisions: FloatArray
        arms = n_arms if n_arms is not None else n_bandits
        if arms is None:
            raise TypeError("n_arms is required")
        super().__init__(n_arms=int(arms), initial_value=prior_mean)

    def reset(self) -> None:
        super().reset()
        self.means = np.full(self.n_arms, self.prior_mean, dtype=float)
        self.precisions = np.full(self.n_arms, self.prior_precision, dtype=float)
        self.estimates = self.means.copy()

    def select_action(self, *, rng: np.random.Generator) -> int:
        z = NormalDist().inv_cdf(self.quantile)
        upper_bounds = self.means + z / np.sqrt(self.precisions)
        return int(np.argmax(upper_bounds))

    def update(self, *, action: int, reward: float) -> None:
        self._validate_action(action)
        self.step += 1
        self.total_reward += float(reward)
        self.counts[action] += 1.0
        precision = self.precisions[action] + self.reward_precision
        mean = (
            self.precisions[action] * self.means[action]
            + self.reward_precision * float(reward)
        ) / precision
        self.precisions[action] = precision
        self.means[action] = mean
        self.estimates[action] = mean

    def __repr__(self) -> str:
        return (
            "GaussianBayesianUCBPolicy("
            f"quantile={self.quantile}, prior_mean={self.prior_mean}, "
            f"reward_precision={self.reward_precision})"
        )


def BayesianUCBPolicy(
    n_arms: int | None = None,
    n_bandits: int | None = None,
    *,
    reward_distribution: str = "gaussian",
    alpha_prior: float = 1.0,
    beta_prior: float = 1.0,
    prior_mean: float = 0.0,
    prior_precision: float = 1.0,
    reward_precision: float = 1.0,
    quantile: float = 0.95,
    **_: object,
) -> Policy:
    """Create a distribution-specific Bayesian UCB policy."""

    arms = n_arms if n_arms is not None else n_bandits
    if arms is None:
        raise TypeError("n_arms is required")
    if reward_distribution == "bernoulli":
        return BernoulliBayesianUCBPolicy(
            n_arms=int(arms),
            alpha_prior=alpha_prior,
            beta_prior=beta_prior,
            quantile=quantile,
        )
    if reward_distribution in {"gaussian", "normal"}:
        return GaussianBayesianUCBPolicy(
            n_arms=int(arms),
            prior_mean=prior_mean,
            prior_precision=prior_precision,
            reward_precision=reward_precision,
            quantile=quantile,
        )
    raise ValueError(f"{reward_distribution!r} cannot be used with Bayesian UCB")

File: crates/pymab-core/src/policy/bayesian_ucb.rs
//! Bayesian upper-confidence-bound policies.

use std::mem::size_of;

use statrs::distribution::{Beta, ContinuousCDF, Normal};

use super::action_value::deterministic_argmax;
use super::thompson::{BernoulliPosteriorState, GaussianPosteriorState};
use super::Policy;
use crate::error::{PyMabError, Result};
use crate::rng::NativeRng;
use crate::types::{
    ActionIndex, PolicyCapabilities, PolicyObjective, RewardDomain, ALL_REWARD_DOMAINS,
};
use crate::validation::strictly_positive;

const BINARY_DOMAINS: &[RewardDomain] = &[RewardDomain::Binary];
const BINARY_CAPABILITIES: PolicyCapabilities =
    PolicyCapabilities::new(false, BINARY_DOMAINS, PolicyObjective::CumulativeReward);
const GENERAL_CAPABILITIES: PolicyCapabilities =
    PolicyCapabilities::new(false, ALL_REWARD_DOMAINS, PolicyObjective::CumulativeReward);

fn open_probability(name: &str, value: f64) -> Result<f64> {
    if value.is_finite() && value > 0.0 && value < 1.0 {
        Ok(value)
    } else {
        Err(PyMabError::configuration(
            name,
            "must be finite and strictly between zero and one",
        ))
    }
}

/// Bayesian UCB using Beta posterior quantiles.
#[derive(Clone, Debug, PartialEq)]
pub struct BernoulliBayesianUCBPolicy {
    alpha_prior: f64,
    beta_prior: f64,
    quantile: f64,
    state: BernoulliPosteriorState,
}

impl BernoulliBayesianUCBPolicy {
    /// Construct a Beta-Bernoulli Bayesian-UCB policy.
    pub fn new(n_arms: usize, alpha_prior: f64, beta_prior: f64, quantile: f64) -> Result<Self> {
        Ok(Self {
            alpha_prior: strictly_positive("alpha_prior", alpha_prior)?,
            beta_prior: strictly_positive("beta_prior", beta_prior)?,
            quantile: open_probability("quantile", quantile)?,
            state: BernoulliPosteriorState::new(n_arms)?,
        })
    }

    /// Return Beta posterior upper bounds for every arm.
    pub fn upper_bounds(&self) -> Result<Vec<f64>> {
        self.state
            .successes()
            .iter()
            .zip(self.state.failures())
            .map(|(successes, failures)| {
                let distribution = Beta::new(
                    self.alpha_prior + *successes as f64,
                    self.beta_prior + *failures as f64,
                )
                .map_err(|error| PyMabError::numerical("Beta quantile", error.to_string()))?;
                let bound = distribution.inverse_cdf(self.quantile);
                if bound.is_finite() {
                    Ok(bound)
                } else {
                    Err(PyMabError::numerical(
                        "Beta quantile",
                        "quantile became non-finite",
                    ))
                }
            })
            .collect()
    }
}

impl Policy for BernoulliBayesianUCBPolicy {
    type State = BernoulliPosteriorState;

    fn capabilities(&self) -> PolicyCapabilities {
        BINARY_CAPABILITIES
    }

    fn n_arms(&self) -> usize {
        self.state.successes().len()
    }

    fn select_action(&mut self, _rng: &mut NativeRng) -> Result<ActionIndex> {
        deterministic_argmax(&self.upper_bounds()?)
    }

    fn update(&mut self, action: ActionIndex, reward: f64) -> Result<()> {
        self.state.update(action, reward)
    }

    fn recommend_action(&self) -> Result<ActionIndex> {
        self.state.action_values().recommendation()
    }

    fn reset(&mut self) {
        self.state.reset();
    }

    fn state(&self) -> &Self::State {
        &self.state
    }

    fn estimated_state_bytes(&self) -> usize {
        size_of::<Self>() + self.state.estimated_heap_bytes()
    }
}

/// Bayesian UCB for Gaussian rewards with known observation precision.
#[derive(Clone, Debug, PartialEq)]
pub struct GaussianBayesianUCBPolicy {
    reward_precision: f64,
    quantile: f64,
    state: GaussianPosteriorState,
}

impl GaussianBayesianUCBPolicy {
    /// Construct a conjugate Gaussian Bayesian-UCB policy.
    pub fn new(
        n_arms: usize,
        prior_mean: f64,
        prior_precision: f64,
        reward_precision: f64,
        quantile: f64,
    ) -> Result<Self> {
        Ok(Self {
            reward_precision: strictly_positive("reward_precision", reward_precision)?,
            quantile: open_probability("quantile", quantile)?,
            state: GaussianPosteriorState::new(n_arms, prior_mean, prior_precision)?,
        })
    }

    /// Return Gaussian posterior upper bounds for every arm.
    pub fn upper_bounds(&self) -> Result<Vec<f64>> {
        let standard_normal =
            Normal::new(0.0, 1.0).map_err(|error| PyMabError::internal(error.to_string()))?;
        let z = standard_normal.inverse_cdf(self.quantile);
        if !z.is_finite() {
            return Err(PyMabError::numerical(
                "Normal quantile",
                "quantile became non-finite",
            ));
        }
        Ok(self
            .state
            .means()
            .iter()
            .zip(self.state.precisions())
            .map(|(mean, precision)| mean + z / precision.sqrt())
            .collect())
    }
}

impl Policy for GaussianBayesianUCBPolicy {
    type State = GaussianPosteriorState;

    fn capabilities(&self) -> PolicyCapabilities {
        GENERAL_CAPABILITIES
    }

    fn n_arms(&self) -> usize {
        self.state.means().len()
    }

    fn select_action(&mut self, _rng: &mut NativeRng) -> Result<ActionIndex> {
        deterministic_argmax(&self.upper_bounds()?)
    }

    fn update(&mut self, action: ActionIndex, reward: f64) -> Result<()> {
        self.state.update(action, reward, self.reward_precision)
    }

    fn recommend_action(&self) -> Result<ActionIndex> {
        self.state.action_values().recommendation()
    }

    fn reset(&mut self) {
        self.state.reset();
    }

    fn state(&self) -> &Self::State {
        &self.state
    }

    fn estimated_state_bytes(&self) -> usize {
        size_of::<Self>() + self.state.estimated_heap_bytes()
    }
}
